packages/repository/src/define-repository-class.ts

```diff
@@ -34,62 +34,73 @@ export interface CrudRepositoryClass<
   prototype: R;
 }
 
-/**
- * Signature for EntityCrudRepository classes
- *
- * @typeParam E - An entity class
- * @typeParam IdType - ID type for the entity
- * @typeParam Relations - Relations for the entity
- */
-export interface EntityCrudRepositoryClass<
-  E extends Entity,
-  IdType,
-  Relations extends object
-> extends CrudRepositoryClass<E, EntityCrudRepository<E, IdType, Relations>> {}
-
 /**
  * Signature for repository classes that can be used as the base class for
- * `define*` functions
+ * `define*` functions. The constructor of a base repository class accepts
+ * the target model constructor and the datasource to use.
+ *
+ * `define*` functions require a class implementing this interface on input.
+ *
+ * @typeParam M - Model class constructor, e.g `typeof Model`.
+ * **❗️IMPORTANT: The type argument `M` is describing the model constructor type
+ * (e.g. `typeof Model`), not the model instance type (`Model`) as is the case
+ * in other repository-related types. The constructor type is required
+ * to support custom repository classes requiring a Model subclass in the
+ * constructor arguments, e.g. `Entity` or a user-provided model.**
  *
- * @typeParam M - Model class
  * @typeParam R - Repository class/interface
  */
-export interface BaseRepositoryClass<M extends Model, R extends Repository<M>> {
+export interface BaseRepositoryClass<
+  M extends typeof Model,
+  R extends Repository<PrototypeOf<M>>
+> {
   /**
    * The constructor for the generated repository class
    * @param modelClass - Model class
    * @param dataSource - DataSource object
    */
-  new (
-    // Make model class conditional of Model or Entity
-    modelClass: (M extends Entity ? typeof Entity : typeof Model) & {
```

Comment:
While this works for Entities, it won't work for other subclasses that people may use for their base repository classes.

Consider the following:

```ts
class MyAuditedEntityRepository<E extends AuditedEntity, /*...*/> extends DefaultCrudRepository<E, /*...*/> {
  constructor(modelClass: typeof AuditedEntity, dataSource: juggler.dataSource) {
    // ...
  }
  // ...
}
```

To support such use case, I think it's best to use `M extends typeof Model` as the generic type parameter.